processor/schemaprocessor/internal/translation/manager.go

```diff
@@ -0,0 +1,139 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package translation // import "github.com/open-telemetry/opentelemetry-collector-contrib/processor/schemaprocessor/internal/translation"
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+)
+
+var errNilValueProvided = errors.New("nil value provided")
+
+// Manager is responsible for ensuring that schemas are kept up to date
+// with the most recent version that are requested.
+type Manager interface {
+	// RequestTranslation will provide either the defined Translation
+	// if it is a known target and able to retrieve from Provider
+	// otherwise it will return an error.
+	RequestTranslation(ctx context.Context, schemaURL string) (Translation, error)
+
+	// AddProvider will add a provider to the Manager
+	AddProvider(p Provider)
+}
+
+type manager struct {
+	log *zap.Logger
+
+	rw            sync.RWMutex
+	providers     []Provider
+	match         map[string]*Version
+	translatorMap map[string]*translator
+}
+
+var _ Manager = (*manager)(nil)
+
+// NewManager creates a manager that will allow for management
+// of schema
+func NewManager(targetSchemaURLS []string, log *zap.Logger, providers ...Provider) (Manager, error) {
+	if log == nil {
+		return nil, fmt.Errorf("logger: %w", errNilValueProvided)
+	}
+
+	match := make(map[string]*Version, len(targetSchemaURLS))
+	for _, target := range targetSchemaURLS {
+		family, version, err := GetFamilyAndVersion(target)
+		if err != nil {
+			return nil, err
+		}
+		match[family] = version
+	}
+
+	// wrap provider with cacheable provider
+	var prs []Provider
+	for _, p := range providers {
+		// TODO make cache configurable
+		prs = append(prs, NewCacheableProvider(p, 5*time.Minute, 5))
+	}
+
+	return &manager{
+		log:           log,
+		match:         match,
+		translatorMap: make(map[string]*translator),
+		providers:     prs,
+	}, nil
+}
+
+func (m *manager) RequestTranslation(ctx context.Context, schemaURL string) (Translation, error) {
+	family, version, err := GetFamilyAndVersion(schemaURL)
+	if err != nil {
+		m.log.Error("No valid schema url was provided, using no-op schema",
+			zap.String("schema-url", schemaURL),
+		)
+		return nil, err
+	}
+
+	targetTranslation, match := m.match[family]
+	if !match {
+		m.log.Warn("Not a known targetTranslation, providing Nop Translation",
```

Comment:
The message seems incorrect, we don't return Nop Translation.

Reply:
good catch. fixed it